confdgnmi/src/confd_gnmi_api_adapter.py

```diff
@@ -59,32 +63,183 @@ def get_inst(cls) -> GnmiServerAdapter:
+            if len(xplist) == 2:
+                xpath = xplist[1]
+                if starts_slash:
+                    xpath = '/' + xpath
+            log.debug("<== xpath=%s", xpath)
+            return xpath
+
+        def process_subscription(self, sub_sock, sub_point):
+            log.debug("==>")
+
+            def iter(kp, op, oldv, newv, state):
+                log.debug("==> kp=%s, op=%r, oldv=%s, newv=%s, state=%r", kp,
+                          op, oldv, newv, state)
+                if op == _confd.MOP_CREATED:
+                    log.debug("_confd.MOP_CREATED")
+                    # TODO CREATE not handled for now
+                if op == _confd.MOP_VALUE_SET:
+                    log.debug("_confd.MOP_VALUE_SET")
+                    self.change_db_lock.acquire()
+                    xpath = self.kp_to_xpath(kp)
+                    self.change_db.append((xpath, str(newv)))
+                    self.change_db_lock.release()
+                    # TODO MOP_VALUE_SET implement
+                elif op == _confd.MOP_DELETED:
+                    log.debug("_confd.MOP_DELETED")
+                    # TODO DELETE not handled for now
+                elif op == _confd.MOP_MODIFIED:
+                    log.debug("_confd.MOP_MODIFIED")
+                    # TODO MODIFIED not handled for now
+                else:
+                    log.warning(
+                        "Operation op=%d is not expected, kp=%s. Skipping!"
+                        , op, kp)
+                return _confd.ITER_RECURSE
+
+            cdb.diff_iterate(sub_sock, sub_point, iter, 0, None)
+            self.put_event(self.SubscriptionEvent.SEND_CHANGES)
+            log.debug("self.change_db=%s", self.change_db)
+            log.debug("<==")
+
+        def process_changes(self):
+            log.debug("==>")
+            # make subscription for all self.monitored_paths
+            sub_sock = socket(AF_INET, SOCK_STREAM, 0)
+            prio = 10
+
+            cdb.connect(sub_sock, cdb.SUBSCRIPTION_SOCKET, '127.0.0.1',
+                        _confd.CONFD_PORT)
+            for p in self.monitored_paths:
+                log.debug("subscribing config p=%s", p)
+                # TODO hash - breaks generic usage
+                # TODO for now we subscribe path for both, config and oper,
+                # TODO subscribe only for paths that exist
+                # it may be more efficient to find out type of path and subscribe
+                # only for one type
+                cdb.subscribe2(sub_sock, cdb.SUB_RUNNING, 0, prio,
+                               0, p)
+                log.debug("subscribing operational p=%s", p)
+                cdb.subscribe2(sub_sock, cdb.SUB_OPERATIONAL, 0, prio,
+                               0, p)
+            cdb.subscribe_done(sub_sock)
+            log.debug("subscribe_done")
+            assert self.stop_pipe is not None
+            _r = [sub_sock, self.stop_pipe[0]]
```

Comment:
That's not a fortunate choice of a variable names, `_r` indicates (at least to me) that its value is not interesting; or is there a Pythonic convention for such names?

Reply:
Yes, it is based on style used in example. I'll rename.